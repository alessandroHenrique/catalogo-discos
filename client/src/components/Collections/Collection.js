import React from 'react';
import PropTypes from 'prop-types';
import DiscList from '../Discs';
import AddDiscForm from '../DiscForm';

class Collection extends React.Component {
  state = {
    isEditing: false,
    textInput: this.props.title,
    searchInput: '',
    discs: this.props.discs || []
  }

  static propTypes = {
    title: PropTypes.string.isRequired,
    discs: PropTypes.array,
    updateCollection: PropTypes.func.isRequired,
    deleteCollection: PropTypes.func.isRequired,
    collectionId: PropTypes.number.isRequired
  }

  textInputChange = (e) => {
    const value = e.target.value;

    this.setState(() => ({
      textInput: value
    }));
  };

  changeIsEditing = (e) => {
    if (e.target.innerHTML === 'Save') {
      this.props.updateCollection(this.state.textInput);
    }
    this.setState((prevState) => ({
      isEditing: !prevState.isEditing
    }));
  }

  searchInputChange = (e) => {
    const value = e.target.value;

    this.setState(() => ({
      searchInput: value
    }));

    this.getVisibleDiscs(value);
  };

  getVisibleDiscs = (text) => {
    this.setState(() => ({
      discs: this.props.discs.filter(
        (disc) => disc.title.includes(text) || disc.author.includes(text) ? disc : false
      )
    }));
  }

  render() {
    const widgetMessage = () => {
      if (this.state.discs.length > 0) {
        return <DiscList discs={this.state.discs} />;
      } else {
        if (this.props.discs && this.props.discs.length > 0) {
          return <span className="widget__message">Nenhum disco encontrado!</span>;
        }
        return <span className="widget__message">Adicione algum disco!</span>;
      }
    }

    return (
      <li className="mt-5 text-center px-0">
        <div className="card widget">
          <div className="card-header widget-header d-flex">
            <h4 className="mr-auto">
              Coleção: {this.state.isEditing ? <input className="add-option__input" type="text" onChange={this.textInputChange} value={this.state.textInput} /> : <span>{this.state.textInput}</span>}
              <button disabled={!this.state.textInput} onClick={this.changeIsEditing} className={`btn ml-2 mr-2 ${this.state.textInput ? 'btn-primary' : 'btn-secondary' }`}>{this.state.isEditing ? 'Save' : 'Edit'}</button>
              <button className="btn btn-danger" onClick={() => this.props.deleteCollection()}>Delete</button>
            </h4>
            <input className="add-option__input search" type="text" onChange={this.searchInputChange} value={this.state.searchInput} placeholder="Search Disc"/>
            <i className="fa fa-search fa-2x" />
          </div>
          <div className="card-body">
            {
              widgetMessage()
            }
          </div>
          <AddDiscForm collectionId={this.props.collectionId} />
        </div>
      </li>
    );
  }
}

export default Collection;
