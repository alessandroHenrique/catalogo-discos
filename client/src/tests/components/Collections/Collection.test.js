import Collection from '../../../components/Collections/Collection';
import React from 'react';
import Store from '../../../store/configureStore';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';

const store = Store();

describe('Collection', () => {
  it('should render a Collection', () => {

    const component = renderer.create(
      <Provider store={store}>
        <Collection
          title={'title'}
          discs={[]}
          deleteCollection={() => {}}
          updateCollection={() => {}}
          collectionId={1}
        />
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })
})
