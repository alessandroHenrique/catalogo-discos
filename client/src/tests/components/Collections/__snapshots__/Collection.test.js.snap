// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection should render a Collection 1`] = `
<li
  className="mt-5 text-center px-0"
>
  <div
    className="card widget"
  >
    <div
      className="card-header widget-header d-flex"
    >
      <h4
        className="mr-auto"
      >
        Coleção: 
        <span>
          title
        </span>
        <button
          className="btn ml-2 mr-2 btn-primary"
          disabled={false}
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          className="btn btn-danger"
          onClick={[Function]}
        >
          Delete
        </button>
      </h4>
      <input
        className="add-option__input search"
        onChange={[Function]}
        placeholder="Search Disc"
        type="text"
        value=""
      />
      <i
        className="fa fa-search fa-2x"
      />
    </div>
    <div
      className="card-body"
    >
      <span
        className="widget__message"
      >
        Adicione algum disco!
      </span>
    </div>
    <form
      className="add-option text-left align-items-center add-disc-form justify-content-center"
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="title"
        >
          Título
        </label>
        <input
          className="add-option__input"
          id="title"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="author"
        >
          Autor
        </label>
        <input
          className="add-option__input"
          id="author"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <button
        className="button"
        type="submit"
      >
        Add Disc
      </button>
    </form>
  </div>
</li>
`;
